File: apps/android/app/src/test/java/dev/crossbridge/android/ConnectionManagerTest.kt
package dev.crossbridge.android

import dev.crossbridge.android.network.applyTrustedDeviceOffline
import dev.crossbridge.android.network.applyTrustedDeviceOnline
import dev.crossbridge.android.network.applyTrustedDeviceStatus
import dev.crossbridge.android.network.createTrustedDeviceConnections
import dev.crossbridge.android.network.reconnectDelayMs
import dev.crossbridge.android.protocol.DeviceIdentity
import dev.crossbridge.android.protocol.TrustedDevice
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ConnectionManagerTest {
    @Test
    fun trustedDevicesStartOffline() {
        val connections = createTrustedDeviceConnections(listOf(windowsDevice()))

        assertEquals(1, connections.size)
        assertFalse(connections.single().online)
        assertEquals(1_500L, connections.single().lastSeenAt)
    }

    @Test
    fun onlineStatusOnlyUpdatesMatchingTrustedDevice() {
        val connections = createTrustedDeviceConnections(listOf(windowsDevice()))
        val online = applyTrustedDeviceOnline(
            devices = connections,
            deviceIdentity = DeviceIdentity(
                deviceId = "pc_1",
                deviceName = "Windows Laptop",
                platform = "windows",
                publicKey = "pc_public_key"
            ),
            timestamp = 2_000L
        )

        assertTrue(online.single().online)
        assertEquals("relay", online.single().connectionMode)
        assertEquals("Windows Laptop", online.single().device.deviceName)
        assertEquals(2_000L, online.single().lastSeenAt)

        val unchanged = applyTrustedDeviceOnline(
            devices = connections,
            deviceIdentity = DeviceIdentity(
                deviceId = "pc_unknown",
                deviceName = "Unknown",
                platform = "windows",
                publicKey = "unknown_public_key"
            ),
            timestamp = 3_000L
        )
        assertEquals(connections, unchanged)
    }

    @Test
    fun onlineStatusSupportsLanFastPath() {
        val connections = createTrustedDeviceConnections(listOf(windowsDevice()))
        val onlineWithLan = connections.map {
            it.copy(localFastPathAvailable = true)
        }
        val online = applyTrustedDeviceOnline(
            devices = onlineWithLan,
            deviceIdentity = DeviceIdentity(
                deviceId = "pc_1",
                deviceName = "Windows PC",
                platform = "windows",
                publicKey = "pc_public_key"
            ),
            timestamp = 2_000L
        )

        assertTrue(online.single().online)
        assertEquals("lan", online.single().connectionMode)
        assertTrue(online.single().localFastPathAvailable)
    }


    @Test
    fun offlineAndStatusUpdatesRefreshLastSeen() {
        val online = applyTrustedDeviceOnline(
            devices = createTrustedDeviceConnections(listOf(windowsDevice())),
            deviceIdentity = DeviceIdentity(
                deviceId = "pc_1",
                deviceName = "Windows PC",
                platform = "windows",
                publicKey = "pc_public_key"
            ),
            timestamp = 2_000L
        )

        val offline = applyTrustedDeviceOffline(online, "pc_1", 3_000L)
        assertFalse(offline.single().online)
        assertEquals(3_000L, offline.single().lastSeenAt)

        val status = applyTrustedDeviceStatus(offline, "pc_1", online = true, lastSeenAt = 4_000L)
        assertTrue(status.single().online)
        assertEquals("relay", status.single().connectionMode)
        assertEquals(4_000L, status.single().lastSeenAt)
    }

    @Test
    fun reconnectBackoffUsesBoundedDelays() {
        assertEquals(
            listOf(1_000L, 2_000L, 5_000L, 10_000L, 30_000L, 30_000L),
            listOf(1, 2, 3, 4, 5, 12).map(::reconnectDelayMs)
        )
    }

    @Test
    fun connectionViewStateHasEmptyTransfers() {
        val state = dev.crossbridge.android.network.ConnectionViewState()
        assertEquals(emptyList<dev.crossbridge.android.network.FileTransferState>(), state.transfers)
    }

    private fun windowsDevice(): TrustedDevice {
        return TrustedDevice(
            deviceId = "pc_1",
            deviceName = "Windows PC",
            platform = "windows",
            publicKey = "pc_public_key",
            pairedAt = 1_000L,
            lastSeenAt = 1_500L
        )
    }
}

File: apps/windows/src/services/connectionManager.test.ts
import { describe, expect, it } from "vitest";
import {
  ConnectionManager,
  applyTrustedDeviceOffline,
  applyTrustedDeviceOnline,
  applyTrustedDeviceStatus,
  createTrustedDeviceConnections
} from "./connectionManager.js";
import { reconnectDelayMs } from "./relayClient.js";
import type { DeviceIdentity, TrustedDevice } from "@crossbridge/protocol";

const pixel: TrustedDevice = {
  deviceId: "android_pixel",
  deviceName: "Pixel",
  platform: "android",
  publicKey: "android_public_key",
  pairedAt: 1_000,
  lastSeenAt: 1_500
};

const updatedPixelIdentity: DeviceIdentity = {
  deviceId: "android_pixel",
  deviceName: "Pixel 8",
  platform: "android",
  publicKey: "android_public_key"
};

describe("connection manager helpers", () => {
  it("creates offline status entries from trusted devices", () => {
    expect(createTrustedDeviceConnections([pixel])).toEqual([
      {
        device: pixel,
        online: false,
        lastSeenAt: pixel.lastSeenAt
      }
    ]);
  });

  it("marks only trusted matching devices online", () => {
    const connections = createTrustedDeviceConnections([pixel]);
    const next = applyTrustedDeviceOnline(connections, updatedPixelIdentity, 2_000);

    expect(next).toEqual([
      {
        device: {
          ...pixel,
          deviceName: "Pixel 8",
          lastSeenAt: 2_000
        },
        online: true,
        connectionMode: "relay",
        lastSeenAt: 2_000
      }
    ]);

    expect(applyTrustedDeviceOnline(connections, {
      ...updatedPixelIdentity,
      deviceId: "android_unknown"
    }, 3_000)).toEqual(connections);
  });

  it("marks trusted devices offline and applies status updates", () => {
    const online = applyTrustedDeviceOnline(
      createTrustedDeviceConnections([pixel]),
      updatedPixelIdentity,
      2_000
    );

    expect(applyTrustedDeviceOffline(online, pixel.deviceId, 3_000)[0]).toMatchObject({
      online: false,
      connectionMode: undefined,
      lastSeenAt: 3_000
    });

    expect(applyTrustedDeviceStatus(online, pixel.deviceId, false, 4_000)[0]).toMatchObject({
      online: false,
      connectionMode: undefined,
      lastSeenAt: 4_000
    });
  });

  it("uses bounded reconnect backoff delays", () => {
    expect([1, 2, 3, 4, 5, 12].map(reconnectDelayMs)).toEqual([
      1_000,
      2_000,
      5_000,
      10_000,
      30_000,
      30_000
    ]);
  });
});

describe("ConnectionManager file transfers", () => {
  it("initializes empty transfers list", () => {
    const mgr = new ConnectionManager();
    expect(mgr.getState().transfers).toEqual([]);
  });

  it("throws offline error when trying to send a file offer to an offline device", async () => {
    const mgr = new ConnectionManager();
    const bytes = new TextEncoder().encode("file content");
    await expect(mgr.sendFileOffer("android_pixel", "notes.txt", "text/plain", bytes))
      .rejects.toThrow(/device is not trusted yet/);
  });
});

